### PR TITLE
interactive console using npyscreen

### PR DESCRIPTION
**Pull request by fhdk**
https://github.com/manjaro/pacman-mirrors/issues/26
Created a tty version. It should launch with `--interactive` if `DISPLAY is None` <strike>but can be forced with `--tty`</strike>

It is not required to select any mirrors presented since an empty list returned results in no changes.
The exchange of lists is file-based with json-data.
If you would review please, the changes, and add your comments.

UPDATE screenshot - Danish language strings is missing - waiting for the po files
console-ui

### CODE REVIEW

**Comment by Huluti** [comment]
For now, I see some problems:
- npyscreen isn't in Manjaro and Arch repos.
- npyscreen is not developped since 2015.
- the strings could be merged with all the interfaces for translation.

**Comment by fhdk** [comment]
As stated earlier my python experience is limited but there is no need to invent the wheel when a freely available library can solve the challenge. So why does have to be in manjaro or arch repo?

It is a python library, available from official python web [https://pypi.python.org/pypi/npyscreen/4.10.5](https://pypi.python.org/pypi/npyscreen/4.10.5). For the developing cycle - I don't see that as obsolete - maybe maintenance is not needed - because it solves its purpose and there is no need for expansion?

Also it exist here [on Github](https://github.com/npcole/npyscreen) and here [in AUR](http://aur.archlinux.org/packages/npyscreen-git) so it is possible to ensure its survival - so to speak.

The strings should be merged, I agree. I didn't do it cause I wanted your opinion on a related matter first. I had the idea to move all strings to a separate file and importing it. Do you reckon it is a bad idea?

**Comment by Huluti** [comment]
When we make a package, we have to add dependencies and the fact is that npyscreen is not in our repos. We couldn't add it as a dependency. Is it right @Kirek?
I like the harmonization given by the texts.py file.
Can you explain me why you use a JSON file (don't enough time to read all changes for now)?

**Comment by philmmanjaro** [comment]
If we need a pkg which is not yet in our repositories, we have to add it there. So do we use **npyscreen** as a dep or as a builddep?

**Comment by Huluti** [comment]
Just as a dep.

**Comment by philmmanjaro** [comment]
@Huluti , @Kirek : do we want to use **npyscreen**? If so we should create that package.

**Comment by Huluti** [comment]
I'm not against.
@fhdk I think that you can re-use a bit more already available strings from the GUI in TTY like the title or the "Confirm" one. It is less work for ours translators and it seems more harmonized.

**Comment by fhdk** [comment]
Are all these libraries part of a default python install?
```
import argparse
import datetime
import importlib.util
import json
import os
import sys
import tempfile
import time
from http.client import HTTPException
from operator import itemgetter
from random import shuffle
from socket import timeout
from urllib.error import URLError
from urllib.request import urlopen
```

I have not quite grasped the passing of variables between modules.

So I used JSON because I like the format and I was able to verify which data was actually exchanged. And the overhead is minimal.

I have looked at the basic module. It seems to me there is no such thing as a selectable list in the basic module - and I think it is partly the reason why others have gone through the effort to create one - and have been kind to share it.

You are more than welcome to show me how to make a selectable list and return the selected items using the basic module. But I think you will rather quick jump to a library. That is why shared libraries exist. Do one thing and do it good.

All texts are shared among modules - There is only one text that does not need translation `OK` :)

**Comment by Huluti** [comment]
I think so. But like @philmmanjaro has said, it's not a problem to add nypscreen in Manjaro repos. So, I understand the use of nypscreen because it's easier to maintain and to write some functionalities. Ok.
The only thing I'm against is the use of two new JSON files. We have to simplify the chain, and keep it simple.
Good work for the texts ;)
Please also think to revert the version, we don't release a new version immediately when a PR is merged.

**Comment by Kirek** [comment]
The are just functions/classes in another file. Just send the stuff you need to send as a parameter and return it with return...

**Comment by Kirek** [comment]
Also now I noticed we are naming all classes PacmanMirrors in all files, better to name them something else, like PacmanMirrorsTui and PacmanMirrorsGui

**Comment by fhdk** [comment]
Tthe json files are not saved in any way - in fact they are created in /tmp and deleted when done.

But I will try the `in-air` exchange :)

Also I will rename classes accordingly - I can't come to an agreement with myself

- TUI or TTY
- the first sounds stupid in my ears
- the last sounds like a remote terminal - which is very close
- also flirted with ConsoleUI

**Comment by fhdk** [comment]
I learned a lot - again :) and I got it to work - also the generation of lines for npyscreen multiselect - so I am happy :). This is the most recent changes in this PR

```
redundant code for interactive removed
staticmethods moved to util.py
texts added
texts modified
test modified to suit new Util class
new potfile generated
linted and corrected accordingly
PacmanMirrors classes renamed
- GraphicalUI - file gui.py
- ConsoleUI - cui.py
```

**Comment by fhdk** [comment]
@Huluti @Kirek 
I have been very thorough and now - if you would be so kind - please review my work

**Comment by Huluti** [comment]
Very great work. I flew over the changes and only one thing comes in my mind for now:Should we rather use basic functions in the file 'util' rather than static methods? I mean not use a class (no-need).

**Comment by fhdk** [comment]
> should we rather use basic functions in the file 'util' rather than static methods? I mean don't use a class (no need)

I don't know the difference but I think separation of concerns is good and I would advise against changing that just for changing - I think you should have a really good argumentation for me to consider that but again - you are free to do what you want - I am just the contributor.

My reasoning is that the methods moved to Util class is self-contained. They don't need the `self` object to fullfill their purpose.

As I see python - it is possible to create code that modifies global variables in a way that is completely confusing. This project contains several parts where the execution logic is not clear unless you are the original developer. I have made an effort to make the code reasonably logic also for people who hasn't seen it before. I would hate to see that wasted.

I have developed large projects in C# and Javascript and my experience is that if a function is doing too much it is time for refactoring into smaller chunks - which in turn are more maintainable. And still there most certainly is room for improvements.

A good example is `self.config["only_country"]` where possible values have nothing to do with a single country. This globals behavior is completely unpredictable. You really have to dive into `load_server_list()` and `config_init()` and `command_line_parse()` because why it has a specific value is absolute not obvious. And I hate when code is not obvious ;) I makes it so hard to understand after a few years - that's is by the way one of my reasons for the heavy commenting when I started contributing to this project.

TL:DR

I think it would be a bad idea :)

**Comment by fhdk** [comment]
There is something bugging me when using -g after creating a custom mirror list.

As is now - the custom mirrorlist is ignored when executing with `-g` even though the `OnlyCountry = Custom` is set in config - It should only be reset by explicit calling with `-c all` so I might have created an unwanted behavior. I will look at it.

**Comment by fhdk** [comment]
@Kirek @Huluti @philmmanjaro 
I have mixed too many changes into this console PR - I will start over and separate the changes into different PRs.

I will keep the Console UI parts based on npyscreen. 

But I am ditching the rest. So for the sake of clarity I will close this PR - delete my branch and make a new branch.

**Comment by Huluti** [comment]
Good idea!
